### PR TITLE
[mlir][flang] Improve handling of fortran module variables.

### PR DESCRIPTION
**Pull request by abidh**
Currently, only those global variables which are at compile unit scope are added to the 'globals' list of the DICompileUnit. This does not work for module variables of Fortran where hierarchy can be variable -> module -> compile unit.

To fix this, if a variable scope points to a module, we walk one level up and see if module is in the compile unit scope.

This was initially part of #91582 which adds debug information for Fortran module variables. @kiranchandramohan pointed out that MLIR changes should go in separate PRs.

### CODE REVIEW

**Comment by Dinistro** [comment]
Can you elaborate on why the variable's scope is not already set accordingly by flang? I assume that there is some reason for this, but I'm not familiar enough with flang to know this detail.

**Comment by abidh** [comment]
> Can you elaborate on why the variable's scope is not already set accordingly by flang? I assume that there is some reason for this, but I'm not familiar enough with flang to know this detail.

Global variable in the context of DWARF is a bit of a loaded term. Same constructs that are used for C/C++ global variables are also used for module variables in Fortran (and namespace variables in C++). MLIR code is assuming that global variable will always have compile unit scope which may not be correct in those cases.

Take the example of namespace in c++ below. The debug metadata was generated by `clang`. You can see that scope of the global variable `a` is set to namespace (and not to compile unit) but it still gets added to `globals` in `DICompileUnit`. 

```
namespace test
{
    int a;
}

!0 = !DIGlobalVariableExpression(var: !1, expr: !DIExpression())
!1 = distinct !DIGlobalVariable(name: "a", linkageName: "_ZN4test1aE", scope: !2 ...)
!2 = !DINamespace(name: "test", scope: null)
!5 = distinct !DICompileUnit(... globals: !6 ...)
!6 = !{!0}
```

Fortran module variables are also represented as global variable in DWARF. But to recognize that they are part of a module, their scope is set to module and not to compile unit. In summary, the scope of the variable is correct. The MLIR check is overly restrictive and this PR amends it for this specific case.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 1]
```diff
@@ -1029,11 +1029,17 @@ LogicalResult ModuleTranslation::convertGlobals() {
           debugTranslation->translateGlobalVariableExpression(op.getDbgExpr());
       llvm::DIGlobalVariable *diGlobalVar = diGlobalExpr->getVariable();
       var->addDebugInfo(diGlobalExpr);
+      // For fortran, the scope hierarchy can be
+      // variable -> module -> compile unit
```
Comment (kiranchandramohan) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1033]: Modules are there in C++ also these days. So can we say "In languages like C++, Fortran etc variables ..."

**Review by kiranchandramohan — COMMENTED** [review]
Thanks @abidh for separating out this patch.

> Take the example of namespace in c++ below. The debug metadata was generated by clang. You can see that scope of the global variable a is set to namespace (and not to compile unit) but it still gets added to globals in DICompileUnit.

Is it possible to have common handling for Namespaces and Modules or will Namespace be another `if` case?

**Comment by abidh** [comment]
> Thanks @abidh for separating out this patch.
> 
> > Take the example of namespace in c++ below. The debug metadata was generated by clang. You can see that scope of the global variable a is set to namespace (and not to compile unit) but it still gets added to globals in DICompileUnit.
> 
> Is it possible to have common handling for Namespaces and Modules or will Namespace be another `if` case?

The namespaces are a bit different as they can be fragmented in multiple files. So their handling may indeed be slightly different if ever a C++ frontend targets mlir.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 1 above)
Comment (kiranchandramohan) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1033]: Or alternatively, 
-> remove the mention of Fortran. Just say that this hierarchy is allowed for languages supporting modules.
-> For Modula 2, Fortran like languages

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Comment by kiranchandramohan** [comment]
> > Thanks @abidh for separating out this patch.
> > > Take the example of namespace in c++ below. The debug metadata was generated by clang. You can see that scope of the global variable a is set to namespace (and not to compile unit) but it still gets added to globals in DICompileUnit.
> > 
> > 
> > Is it possible to have common handling for Namespaces and Modules or will Namespace be another `if` case?
> 
> The namespaces are a bit different as they can be fragmented in multiple files. So their handling may indeed be slightly different if ever a C++ frontend targets mlir.

The clangir project (https://github.com/llvm/clangir) is being upstreamed. https://discourse.llvm.org/t/rfc-upstreaming-clangir/76587

**Comment by Dinistro** [comment]
> Fortran module variables are also represented as global variable in DWARF. But to recognize that they are part of a module, their scope is set to module and not to compile unit. In summary, the scope of the variable is correct. The MLIR check is overly restrictive and this PR amends it for this specific case.

Thanks for the clarification. Ideally, we would model `DICompile` units to also have a `globals` field, but this would result in circular attribute structures. While supporting such constructs is possible, it's quite involved to get this to work and is tricky to work with.

**Review by Dinistro — APPROVED** [review]
LGTM % nit

**Comment by abidh** [comment]
> > Fortran module variables are also represented as global variable in DWARF. But to recognize that they are part of a module, their scope is set to module and not to compile unit. In summary, the scope of the variable is correct. The MLIR check is overly restrictive and this PR amends it for this specific case.
> 
> Thanks for the clarification. Ideally, we would model `DICompile` units to also have a `globals` field, but this would result in circular attribute structures. While supporting such constructs is possible, it's quite involved to get this to work and is tricky to work with.

`DICompileUnit` does have a globals field. Although it is not exposed through DIBuilder::createCompileUnit. Generally, when `DICompileUnit` is created, you don't have the list of globals. The `DICompileUnit` has `replaceGlobalVariables` to set the list of globals once they have been processed, providing a way to break circular dependency.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 1 above)
Comment (abidh) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Thanks for the suggestion. I have updated the comments.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 2]
```diff
@@ -1029,11 +1029,17 @@ LogicalResult ModuleTranslation::convertGlobals() {
...
+      // If a variable scope points to Module then we get its parent scope so
+      // that globals get added to the compile unit.
+      llvm::DIScope *scope = diGlobalVar->getScope();
```
[the remark this answers is not in the document]
Comment (abidh) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1041]: I have updated the comments.

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir [hunk 3]
```diff
@@ -311,6 +311,29 @@ llvm.mlir.global external @global_with_expr_2() {addr_space = 0 : i32, dbg_expr
...
+// CHECK-DAG: ![[GVALS]] = !{![[GEXPR0]], ![[GEXPR1]]}
+
+#di_file = #llvm.di_file<"test.f90" in "existence">
+#di_compile_unit = #llvm.di_compile_unit<id = distinct[0]<>, sourceLanguage = DW_LANG_Fortran95, file = #di_file, producer = "MLIR", isOptimized = true, emissionKind = Full>
+#di_basic_type = #llvm.di_basic_type<tag = DW_TAG_base_type, name = "integer", sizeInBits = 64, encoding = DW_ATE_signed>
+#di_module = #llvm.di_module<file = #di_file, scope = #di_compile_unit, name = "module2", configMacros = "", includePath = "", apinotes = "", line = 120, isDecl = false >
+llvm.mlir.global external @module_global_1() {dbg_expr = #llvm.di_global_variable_expression<var = <scope = #di_module, name = "module_global_1", linkageName = "module_global_1", file = #di_file, line = 121, type = #di_basic_type, isLocalToUnit = false, isDefined = true>, expr = <>>} : i64
+llvm.mlir.global external @module_global_2() {dbg_expr = #llvm.di_global_variable_expression<var = <scope = #di_module, name = "module_global_2", linkageName = "module_global_2", file = #di_file, line = 122, type = #di_basic_type, isLocalToUnit = false, isDefined = true>, expr = <>>} : i64
```
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:333]: Do we require both of these for the test here. They seem identical, and only using this to ensure we fill a list properly seems a bit overkill, given that this is tested in other places already.

Comment (abidh) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir:333]: Yes, one would be enough. I would update it.

Comment (abidh) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir]: Done.

**Comment by Dinistro** [comment]
Thanks for addressing the comments. Feel free to land, btw :)

**Pull request merged**